src/dataflow/src/render/mod.rs

```diff
@@ -674,26 +674,67 @@ pub(crate) fn build_dataflow<A: Allocate>(
                     ))
                     .expect("Sink source collection not loaded");
 
-                // TODO(frank): consolidation is only required for a collection,
-                // not for arrangements. We can perform a more complicated match
-                // here to determine which case we are in to avoid this call.
-                use differential_dataflow::operators::consolidate::Consolidate;
-                let collection = collection.consolidate();
+                // TODO(frank): Consolidation can be avoided for arranged inputs.
+
+                // These steps perform consolidation, but do not devolve the stream
+                // from batches to individual updates. That can happen for sinks that
+                // require e.g. updates shuffled, but for sinks that do not require
+                // such it is much more efficient to work off of shared batches.
+                use differential_dataflow::trace::implementations::ord::{
+                    OrdKeyBatch, OrdKeySpine,
+                };
+                let batches: timely::dataflow::Stream<
+                    _,
+                    std::rc::Rc<OrdKeyBatch<Row, Timestamp, isize>>,
+                > = collection
+                    .map(|k| (k, ()))
+                    .arrange_named::<OrdKeySpine<_, _, _>>("SinkConsolidation")
+                    .stream;
 
                 // TODO(benesch): errors should stream out through the sink,
                 // if we figure out a protocol for that.
 
+                // Common imports for devolving streams of batches.
+                use differential_dataflow::trace::cursor::CursorDebug;
+                use differential_dataflow::trace::BatchReader;
+
                 let sink_shutdown = match sink.connector {
                     SinkConnector::Kafka(c) => {
-                        let button = sink::kafka(&collection.inner, sink_id, c, sink.from.1);
+                        // Devolution code; batches could provide better methods for iteration.
+                        let collection = batches.flat_map(|batch| {
```

Comment:
I know this variable is called a collection but this is actually a stream of (Row, Timestamp, Diff) right?